File: 3-matmult/array/matmult-mpi.c
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"
#include <math.h>

// Valor constante para llenar la matriz
static double a = 1.0E-10;

void FillMatrix(double *matrixA, double *matrixB, int N);
void Multiply(double *matrixA, double *matrixB, double *matrixC, int Nx, int Ny);
double AddMatrix(double *matrix, int Nx, int Ny);

int main(int argc, char *argv[])
{
    int N = 5;             // Dimension de la matriz
    int rank;              // Indice de cada proceso
    int proccess;          // Numero total de procesos
    int waste;             // Residuo de informacion
    int n_local;           // Tamaño de informacion por proceso
    int processSize;       // Tamaño corregido
    int i;                 // Iterador de procesos
    double matrixA[N * N]; // Primera matriz
    double matrixB[N * N]; // Segunda matriz
    double matrixC[N * N]; // Matriz resultado
    double result;         // Resultado de la suma de la matriz resultado
    double result_local;   // Resultado de la suma local de cada proceso
    double estimation;     // Estimacion del calculo
    double error;          // Error encontrado

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &proccess);

    // Calcula el sobrante de datos en caso de que los procesos no sean
    // multiplos de la informacion
    waste = N % (proccess - 1);

    // Calcula el numero de datos que va a tomar cada proceso
    n_local = N / (proccess - 1);

    if (rank == 0)
    {
        // Inicializa el resultado en 0.0
        result = 0.0;

        // Llena la matriz A y B con el valor constante
        FillMatrix(matrixA, matrixB, N);
    }

    // Comparte la informacion de las matrices A y B con los demas procesos
    MPI_Bcast(matrixA, N * N, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    MPI_Bcast(matrixB, N * N, MPI_DOUBLE, 0, MPI_COMM_WORLD);

    if (rank == 0)
    {
        // Recibe el resultado de la suma de todos los elementos de la matriz C
        for (i = 1; i < proccess; i++)
        {
            MPI_Recv(&result_local, 1, MPI_DOUBLE, i, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            result += result_local;
        }

        // Calculo estimado con la formula a^2*N^3.
        estimation = pow(N, 3) * pow(a, 2);

        // Calcula el % de error.
        error = fabs(result - estimation) / estimation * 100.0;

        // Imprime el % de error.
        printf("Error %.15le N = %d\n", error, N);
    }
    else
    {
        // Verifica si es el ultimo proceso para calcular el reciduo
        if (rank == proccess - 1)
            processSize = waste + n_local;
        else
            processSize = n_local;

        // Hace la multiplicacion de matrices.
        Multiply(matrixA, matrixB, matrixC, processSize, N);

        // Hace la suma de todos los elementos de la matriz C
        result_local = AddMatrix(matrixC, processSize, N);

        // Envia el resultado de la multiplicacion al proceso 0.
        MPI_Send(&result_local, 1, MPI_DOUBLE, 0, 1, MPI_COMM_WORLD);

    }

    // Termina MPI
    MPI_Finalize();

    return 0;
}

// Llena las dos matrices con el valor constante
void FillMatrix(
    double *matrixA, // Primera matriz
    double *matrixB, // Segunda matriz
    int N            // Dimension de la matriz
)
{
    int i; // Indice el renglon
    int j; // Indice de la columna

    for (i = 0; i < N; i++)
        for (j = 0; j < N; j++)
        {
            matrixA[(i * N) + j] = a;
            matrixB[(i * N) + j] = a;
        }
}

// Multiplica las dos matrices y almacena el resultado en la matriz de resultado
void Multiply(
    double *matrixA, // Primera matriz
    double *matrixB, // Segunda matriz
    double *matrixC, // Matriz resultado
    int Nx,          // Tamaño de renglones
    int Ny           // Tamaño de columnas
)
{
    int i;         // Indice del renglon
    int j;         // Indice de la columna
    int k;         // Indice de la multiplicacion
    double result; // Resultado de la multiplicacion

    for (i = 0; i < Nx; i++)
        for (j = 0; j < Ny; j++)
        {
            result = 0.0;
            for (k = 0; k < Ny; k++)
                result += matrixA[(i * Ny) + k] * matrixB[(k * Ny) + j];
            matrixC[(i * Ny) + j] = result;
        }
}

// Suma todos los elementos de una matriz y regresa el resultado
double AddMatrix(
    double *matrix, // Matriz resultado
    int Nx,         // Tamaño de renglones
    int Ny          // Tamaño de columnas
)
{
    double result; // Resultado de la suma
    int i;         // Indice del renglon
    int j;         // Indice de la columna

    result = 0.0;

    for (i = 0; i < Nx; i++)
        for (j = 0; j < Ny; j++)
            result += matrix[(i * Nx) + j];

    return result;
}

File: 3-matmult/array/matmult-mpi.f90
! File: matmult-mpi.f90
!
! Purpose: 
! 
! Input:
! 
! Output:
! 
! Compile: mpif90_intel -o matmult-mpi-f90.o matmult-mpi.f90
! 
! Run: mpiexec_intel -np 10 ./matmult-mpi-f90.o
! 
! Algorithm:
! 
! Note:

program matmult
    implicit none
    include 'mpif.h'

    INTERFACE 
    subroutine PrintMatrix(matrix, size, name)
        REAL, INTENT(IN) :: matrix(:)
        INTEGER, INTENT(IN) :: size
        CHARACTER(len=*), INTENT(IN) :: name
    end subroutine PrintMatrix

        subroutine FillMatrix(array, size)
            REAL, INTENT(OUT) :: array(:)
            INTEGER, INTENT(IN) :: size
        end subroutine FillMatrix

        subroutine Multiply(MatrixA, MatrixB, MatrixC, sizeX, sizeY)
            REAL, INTENT(IN) :: MatrixA(:)
            REAL, INTENT(IN) :: MatrixB(:)
            REAL, INTENT(OUT) :: MatrixC(:)
            INTEGER, INTENT(IN) :: sizeX
            INTEGER, INTENT(IN) :: sizeY
        end subroutine Multiply

    END INTERFACE

    INTEGER, PARAMETER :: size = 5
    REAL, DIMENSION(size * size) :: matrixA
    REAL, DIMENSION(size * size) :: matrixB
    REAL, DIMENSION(size * size) :: matrixC
    INTEGER :: rank 
    INTEGER :: process
    INTEGER :: processSize
    INTEGER :: ierror
    INTEGER :: status(MPI_STATUS_SIZE)
    INTEGER :: waste
    INTEGER :: n
    INTEGER :: i

    call MPI_INIT(ierror)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, process, ierror)
    call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierror)

    waste = mod(size, process - 1)
    n = size / (process - 1)

    if (rank == 0) then

        call FillMatrix(matrixA, size)
        call FillMatrix(matrixB, size)

    end if

    call MPI_BCAST(matrixA, size * size, MPI_REAL, 0, MPI_COMM_WORLD, ierror)
    call MPI_BCAST(matrixB, size * size, MPI_REAL, 0, MPI_COMM_WORLD, ierror)

    if (rank == 0) then
        
        processSize = n

        do i = 1, process - 1, 1

            if (i == process - 1) then
                processSize = waste + n
            end if

            call MPI_RECV(matrixC((i - 1) * size * n + 1), processSize * size, MPI_REAL, i, 0, MPI_COMM_WORLD, status, ierror)

        end do

        call PrintMatrix(matrixA, size, 'Matrix A')
        call PrintMatrix(matrixB, size, 'Matrix B')
        call PrintMatrix(matrixC, size, 'Matrix C (result)')

    else 

        if (rank == process - 1) then
            processSize = waste + n;
        else
            processSize = n;
        end if

        call Multiply(matrixA, matrixB, matrixC, processSize, size)
        call MPI_SEND(matrixC, processSize * size, MPI_REAL, 0, 0, MPI_COMM_WORLD, ierror)

    end if

    call MPI_FINALIZE(ierror)

end program matmult

subroutine PrintMatrix(matrix, size, name)
    IMPLICIT NONE
    REAL, INTENT(IN) :: matrix(:)
    INTEGER, INTENT(IN) :: size
    CHARACTER(len=*), INTENT(IN) :: name
    INTEGER :: i
    INTEGER :: j

    write(*, *) name
    do i = 1, size, 1
        do j = 1, size, 1
            write (*,*) i, j, matrix((i - 1) * size + j) 
        end do
    end do

end subroutine PrintMatrix

subroutine FillMatrix(matrix, size)
    IMPLICIT NONE
    REAL, INTENT(OUT) :: matrix(:)
    INTEGER, INTENT(IN) :: size
    INTEGER :: i
    INTEGER :: j

    do i = 1, size, 1
        do j = 1, size, 1
            matrix((i - 1) * size + j)  = 1
        end do
    end do

end subroutine FillMatrix

subroutine Multiply(MatrixA, MatrixB, MatrixC, sizeX, sizeY)
    IMPLICIT NONE
    REAL, INTENT(IN) :: MatrixA(:)
    REAL, INTENT(IN) :: MatrixB(:)
    REAL, INTENT(OUT) :: MatrixC(:)
    INTEGER, INTENT(IN) :: sizeX
    INTEGER, INTENT(IN) :: sizeY
    INTEGER :: i
    INTEGER :: j
    INTEGER :: k
    REAL :: result

    do i = 1, sizeX, 1
        do j = 1, sizeY, 1

            result = 0.0
            do k = 1, sizeY, 1
                result = result + matrixA((i - 1) * sizeY + k) * matrixB((k - 1) * sizeY + j)
            end do
            matrixC((i - 1) * sizeY + j) = result
        
        end do
    end do

end subroutine Multiply